### ISSUE TITLE
Use `allowance` accessor in `transferFrom`

### ISSUE THREAD

**Issue by Amxx** [OpenZeppelin/openzeppelin-contracts#3154]
Currently, transferFrom does

```
function transferFrom(
    address sender,
    address recipient,
    uint256 amount
) public virtual override returns (bool) {
    uint256 currentAllowance = _allowances[sender][_msgSender()];
    if (currentAllowance != type(uint256).max) {
        require(currentAllowance >= amount, "ERC20: transfer amount exceeds allowance");
        unchecked {
            _approve(sender, _msgSender(), currentAllowance - amount);
        }
    }

    _transfer(sender, recipient, amount);

    return true;
}
```

In some cases, you may want to give a privileged contract the right to transfer tokens out of any user. This could for example be a staking contract that the token is aware of. When calling "deposit" on the stacking contract, that would relay a transferFrom. We may want to avoid the user having to set an allowance for that particular contract.

The logic way to do that would be to override allowance
```
function allowance(address owner, address spender) public view virtual override returns (uint256) {
    return spender == staking 
        ? type(uint256).max
        : _allowances[owner][spender];
}
```

This would work well with the latest ERC20 refactor. The allowance being infinite, this would not trigger an `_approve` call, and not emit an `Approval` event.

Unfortunetally, `transferFrom` reads directly from the mapping. IMO `transferFrom` should get `uint256 currentAllowance` by calling the (potentially overridden) `allowance` function.

**Comment by frangio** [comment]
I don't know if this would be a good idea, but if we can do it with no downside I think it's ok.

**Comment by frangio** [comment]
Fixed in #3167.
